From: SJ Park <sj@kernel.org>
Date: Mon, 3 Aug 2026 07:37:50 -0700
Subject: [PATCH] mm/damon/core: error if damon_call() is called for unstarted
 ctx

Since it is cumbersome to check if ctx is started from callers.

Signed-off-by: SJ Park <sj@kernel.org>
---
 mm/damon/core.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index be8e2a0c09fc9..0093f67cf6cee 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2358,16 +2358,14 @@ int damon_kdamond_pid(struct damon_ctx *ctx)
  * synchronization.  The return value of the function will be saved in
  * &damon_call_control->return_code.
  *
- * Note that this function should be called only after damon_start() with the
- * @ctx has succeeded.  Otherwise, this function could fall into an indefinite
- * wait.
- *
  * When this function is failed, the @ctx is guaranteed to be stopped.
  *
  * Return: 0 on success, negative error code otherwise.
  */
 int damon_call(struct damon_ctx *ctx, struct damon_call_control *control)
 {
+	if (!completion_done(ctx->kdamond_started))
+		return -EINVAL;
 	if (!control->repeat)
 		init_completion(&control->completion);
 	control->canceled = false;
-- 
2.47.3

